src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs

```diff
@@ -249,5 +256,42 @@ private static ParameterInfo GetParameterInfo(INamedTypeSymbol type, bool isArra
         {
             return ParameterInfo.GetParameterInfo(type, isArray, arrayRank, isParams);
         }
+
+        private static bool IsValidToStringCall(IInvocationOperation invocationOperation, ImmutableHashSet<INamedTypeSymbol> invariantToStringTypes,
+            INamedTypeSymbol? dateTimeType, INamedTypeSymbol? dateTimeOffsetType, INamedTypeSymbol? timeSpanType)
+        {
+            var targetMethod = invocationOperation.TargetMethod;
+
+            if (targetMethod.Name != "ToString")
+            {
+                return false;
+            }
+
+            if (invariantToStringTypes.Contains(targetMethod.ContainingType))
+            {
+                return true;
+            }
+
+            if (invocationOperation.Arguments.Length != 1 ||
+                !invocationOperation.Arguments[0].Value.ConstantValue.HasValue ||
+                !(invocationOperation.Arguments[0].Value.ConstantValue.Value is string format))
+            {
+                return false;
+            }
+
+            // Handle invariant format specifiers, see https://github.com/dotnet/roslyn-analyzers/issues/3507
+            if ((dateTimeType != null && targetMethod.ContainingType.Equals(dateTimeType)) ||
+                (dateTimeOffsetType != null && targetMethod.ContainingType.Equals(dateTimeOffsetType)))
+            {
+                return format == "o" || format == "O" || format == "r" || format == "R" || format == "s" || format == "u";
```

Comment:
Nit: write it as a switch expression or define a static readonly hashset/array and do a contains check?